test/test_transforms_v2_refactored.py

```diff
@@ -0,0 +1,597 @@
+
+        spy.assert_any_call(f"{dispatcher.__module__}.{dispatcher.__name__}")
+
+    unknown_input = object()
+    with pytest.raises(TypeError, match=re.escape(str(type(unknown_input)))):
+        dispatcher(unknown_input, *args, **kwargs)
+
+    if check_scripted_smoke:
+        _check_dispatcher_scripted_smoke(dispatcher, input, *args, **kwargs)
+
+    if check_dispatch_simple_tensor:
+        _check_dispatcher_dispatch_simple_tensor(dispatcher, kernel, input, *args, **kwargs)
+
+    if check_dispatch_datapoint:
+        _check_dispatcher_dispatch_datapoint(dispatcher, kernel, input, *args, **kwargs)
+
+    if check_dispatch_pil:
+        _check_dispatcher_dispatch_pil(dispatcher, input, *args, **kwargs)
+
+
+def _check_dispatcher_kernel_signature_match(dispatcher, *, kernel, input_type):
+    dispatcher_signature = inspect.signature(dispatcher)
+    dispatcher_params = list(dispatcher_signature.parameters.values())[1:]
+
+    kernel_signature = inspect.signature(kernel)
+    kernel_params = list(kernel_signature.parameters.values())[1:]
+
+    # We filter out metadata that is implicitly passed to the dispatcher through the input datapoint, but has to be
+    # explicit passed to the kernel.
+    kernel_params = [param for param in kernel_params if param.name not in input_type.__annotations__.keys()]
+
+    dispatcher_params = iter(dispatcher_params)
+    for dispatcher_param, kernel_param in zip(dispatcher_params, kernel_params):
+        try:
+            # In general, the dispatcher parameters are a superset of the kernel parameters. Thus, we filter out
+            # dispatcher parameters that have no kernel equivalent while keeping the order intact.
+            while dispatcher_param.name != kernel_param.name:
+                dispatcher_param = next(dispatcher_params)
+        except StopIteration:
+            raise AssertionError(
+                f"Parameter `{kernel_param.name}` of kernel `{kernel.__name__}` "
+                f"has no corresponding parameter on the dispatcher `{dispatcher.__name__}`."
+            ) from None
+
+        assert dispatcher_param == kernel_param
+
+
+def _check_dispatcher_datapoint_signature_match(dispatcher):
+    dispatcher_signature = inspect.signature(dispatcher)
+    dispatcher_params = list(dispatcher_signature.parameters.values())[1:]
+
+    datapoint_method = getattr(datapoints._datapoint.Datapoint, dispatcher.__name__)
+    datapoint_signature = inspect.signature(datapoint_method)
+    datapoint_params = list(datapoint_signature.parameters.values())[1:]
+
+    # Because we use `from __future__ import annotations` inside the module where `datapoints._datapoint` is
+    # defined, the annotations are stored as strings. This makes them concrete again, so they can be compared to the
+    # natively concrete dispatcher annotations.
+    datapoint_annotations = get_type_hints(datapoint_method)
+    for param in datapoint_params:
+        param._annotation = datapoint_annotations[param.name]
```

Comment:
Is it strictly necessary to set the `_annotation` here?
Can we compare the signature without "manually" manipulating annotations?

Reply:
It is not, but IMO it is more convenient. Each parameter carries [four attributes](https://github.com/python/cpython/blob/1858db7cbdbf41aa600c954c15224307bf81a258/Lib/inspect.py#L2697-L2717):

- `name`
- `default`
- `annotation`
- `kind`

We could extract all of them into a tuple and replace the annotation there, but won't make the implementation easier.